### ISSUE TITLE
Improve statistics for secondary hits/misses

### ISSUE THREAD

**Issue by orgads** [ccache/ccache#1016]
This is a sample output of `ccache -s`.
```
Summary:
  Hits:               0 /  675 (0.00 %)
    Direct:           0 /  675 (0.00 %)
    Preprocessed:     0 /  675 (0.00 %)
  Misses:           675
    Direct:         675
    Preprocessed:   675
Primary storage:
  Hits:               0 / 1348 (0.00 %)
  Misses:          1348
  Cache size (GB): 0.37 / 1.00 (37.09 %)
Secondary storage:
  Hits:             675 / 1348 (50.07 %)
  Misses:           673
```

### Actual behavior ###
When I try to analyze this, if I got it correctly, it looks like there were 675 executions which were all found in the secondary storage, but probably preprocessed.

So each preprocessed *hit* is counted as a miss and a hit.

The primary storage had no hits at all (it was cleared before the execution), but the miss count is 1348, which is misleading, since there were actually only 675 executions.

This is very misleading. Notice also the 50% hit ratio on secondary storage, which should actually be 100%, since there were no real misses (only direct).

### Expected behavior ###
Since a missed direct is always followed by preprocessed, I'd expect an execution to be counted as a miss only after preprocessed fails.

You can still count direct misses separately, but the total misses should only be updated after PP miss.

The output on this case should be something like:
```
Summary:
  Hits:               0 /  675 (0.00 %)
    Direct:           0 /  675 (0.00 %)
    Preprocessed:     0 /  675 (0.00 %)
  Misses:           675
    Direct:         675
    Preprocessed:   675
Primary storage:
  Hits:               0 / 675 (0.00 %)
  Misses:          675
  Cache size (GB): 0.37 / 1.00 (37.09 %)
Secondary storage:
  Hits:             675 / 675 (100.00 %)
  Misses:           0
```

### Environment ###
ccache version 4.6

Clang on CentOS 8

**Comment by orgads** [comment]
I also don't understand how can I have 675 secondary hits, but 0/675 in the summary...

**Comment by jrosdahl** [comment]
I don't see anything unexpected given the information you supply.

The _Summary_ section counts hits and misses as in "**did ccache have to run the compiler?**". The "direct hits/misses" parts of this tell how many times ccache tried to compute the hash key via manifest data and the result of doing so (i.e., if ccache tried direct lookup then a direct miss if the compiler did not have to be run, otherwise a direct miss). Similarly, "preprocessed hits/misses" tell how many times ccache had to run the preprocessor in order to compute the hash key and the result of doing so (i.e., if ccache ran the preprocessor then a preprocessed hit if the compiler did not have to be run, otherwise preprocessed miss).

In contrast, the _Primary/Secondary storage_ sections count hits and misses for **lookups in the storage layer**. Unless you have disabled direct or preprocessed mode, there are **two** lookups per compiler invocation: one for the _manifest data structure_ and one for the _result data structure_. The ≈50% hit rate for secondary storage is likely because you got hits for manifest data but no hits for result data.

> When I try to analyze this, if I got it correctly, it looks like there were 675 executions

Right.

> which were all found in the secondary storage, but probably preprocessed.

No, all executions had to run the real compiler, so no matching result data was found. Whether the result key was computed using the preprocessor or not has nothing to do with the storage layer. The storage layer is essentially a simple key-value store which doesn't know or care about how the key was computed.

> So each preprocessed _hit_ is counted as a miss and a hit.

Nope. :slightly_smiling_face:

> The primary storage had no hits at all (it was cleared before the execution), but the miss count is 1348, which is misleading, since there were actually only 675 executions.

I can understand that it might be confusing that hits and misses mean different things in the different sections. If you have suggestions on how to improve the presentation or the way to count statistics I'm all ears.

**Comment by orgads** [comment]
Thank you very much for the detailed explanation.

Can you elaborate a bit about the different lookups?

By manifest do you refer to something like the compiler binary, flags and hash of the file contents?

What's the *value* for the manifest? Does it contain a "pointer" for the key of the actual result? Does it contain an object file? Something else?

**Comment by jrosdahl** [comment]
I hope that the [_How ccache works_](https://ccache.dev/manual/4.6.html#_how_ccache_works) section in the ccache manual and the subsection _The direct mode_ explains this clearly enough.

**Comment by orgads** [comment]
Thanks.

I still don't understand how did I have 675 hits, but still the compiler had to be run for each file? What did ccache find in the cache for these hits?

**Comment by jrosdahl** [comment]
> I still don't understand how did I have 675 hits, but still the compiler had to be run for each file? What did ccache find in the cache for these hits?

Manifest data structures matching the hash of source code file content + compiler options. Those manifest structures (from previous compilations) contain information about _candidate_ results, of which none actually matched. This happens when the source code + compiler options are identical but some header file content has changed.

**Comment by orgads** [comment]
Ok, that's what I called "a pointer". That's practically metadata that is itself stored in the cache, if I understand correctly.

So would it be possible to count the manifest lookups separately?

**Comment by jrosdahl** [comment]
Yes, it would be possible to have separate counters for different cache entry types. However, I consider manifest and result entries to be implementation details, so I'm not sure if it would make sense to expose that in the statistics output. Maybe with `-v` in that case.

By the way, it's possible that there will be third cache entry type in the future if we implement deduplication of file data stored in the cache. (This is a low priority feature on my to do list, and the APIs and architecture have been prepared with that use case in mind.) Then each result entry would contain references to one or several file entries. Each file entry will be a separate lookup over the storage layer. A "cache hit" could then get some files from primary storage and some files from secondary storage (since a subset of the file entries may have been garbage collected from primary storage).
